### DISCUSSION TITLE
Modified mirror sample does not seem to be working

### DISCUSSION THREAD

**Discussion by lostmsu** [dokan-dev/dokan-dotnet#326] [Q&A]
Haven't seen it anywhere, but are breakpoints inside the `IDokanOperations` supposed to be hit?

What I did is slightly modified `Mirror` to attempt to download files and folders from remote source when they don't exist locally. E.g. Mount Point -> Mirror -> Remote Source.

```
options.MountPoint = ...;
// note I skip notifications for simplicity for now
options.Options = DokanOptions.DebugMode;
```

The problem is: although my call to `DokanInstanceBuilder.Build` succeeds and the file system traces

> Debug: Mounted('(null)', {<null>, False, False, False, False, #0, False, False}) -> Success

When I go to the directory specified in `MountPoint`, I don't see contents of the source Mirror directory (it has a single README file). Moreover, my test app then creates a few files under the mount point, and I don't see them in the source Mirror directory at all. Moreover, when the app stops, the test files it created are still accessible under the mount point. Moreover, if I start the app again without removing them first, it fails to assign drive presumably because the mount point is not empty (which is correct, but indicates, that mounted FS is completely ignored).

Neither of my breakpoints hit, and no trace is printed until I close my test app, upon which

> Debug: Unmounted('(null)', {<null>, False, False, False, False, #0, False, False}) -> Success

is printed.

**Comment by Liryna** [comment]
Hi @lostmsu ,

Do you see those logs https://github.com/dokan-dev/dokan-dotnet/blob/master/DokanNet/DokanOperationProxy.cs#L309-L318 ?

Does the default C# samples work ? If not, have you tried with the native sample available in the dokan install location ?
Have you tried with a simple drive letter as mount point ?

**Reply by lostmsu** [reply]
I don't see these lines. There are no logs after Mounted message until I close the app when it prints Unmounted.

Will try the original sample itself once I get home.

**Reply by lostmsu** [reply]
Sigh. The original sample appears to be working as intended.

Could the silent failure to work be a sign of an unhandled exception?

**Reply by lostmsu** [reply]
OK, I believe part of the issue is with trying to access the mount point too early or something like that.

My test app originally looked like this:
```csharp
using (var savesInstance = savesBuilder.Build(saves)) {
    logger.LogDebug("Dokan started");
    Debug.Assert(savesInstance.IsFileSystemRunning());
    using var saveGenerator = new SaveGenerator();
    saveGenerator.RunGenerateSavesLoop();
}
```

Now I added 3 minute `Thread.Sleep` before `SaveGenerator`, and the contents of the folder show up in Explorer as expected, including the "remote" bits. And I face a reproducible issue with trying to open a file in Notepad, edit, save, close Notepad and open the file again - this fails, and Notepad hangs, and I get

> DummyGame Error: 0 : ReadFileProxy : \TEST.txt Throw : Input string was not in a correct format.

Is it possible to see where this error was raised? Or even better hit it in the debugger?

**Reply by Olof-Lagerkvist** [reply]
Yes you really need to find out where that issue happens. You could enable break on all exceptions in exception settings in Visual Studio or you can put a breakpoint on the catch statements in the proxy class methods.

**Reply by lostmsu** [reply]
@LTRData tried that, not breaking into the debugger.

**Reply by lostmsu** [reply]
OK, got the breakpoint in the sources decompiled by Rider (oh I wish there was a better way to debug external packages). Now it looks to me like a bug in Dokan.  This is the relevant bit of the stack trace:

>    at System.String.Format(IFormatProvider provider, String format, Object[] args)
>    at System.Diagnostics.TraceListener.TraceEvent(TraceEventCache eventCache, String source, TraceEventType eventType, Int32 id, String format, Object[] args)
>    at System.Diagnostics.TraceInternal.TraceEvent(TraceEventType eventType, Int32 id, String format, Object[] args)
>    at System.Diagnostics.Trace.TraceInformation(String format, Object[] args)
>    at DokanNet.Logging.TraceLogger.Debug(String message, Object[] args)
>    at DokanNet.DokanOperationProxy.ReadFileProxy(String rawFileName, IntPtr rawBuffer, UInt32 rawBufferLength, Int32& rawReadLength, Int64 rawOffset, DokanFileInfo rawFileInfo)

Not sure if the inability of the VS and Rider to break on that exception on their own is a bug or not, but logging in `ReadFileProxy` definitely should not be rising `FormatException`.

**Reply by lostmsu** [reply]
Opened https://github.com/dokan-dev/dokan-dotnet/pull/327

**Reply by Olof-Lagerkvist** [reply]
Ah, you mean you have Dokan as nuget dependency package and not a debug build? Then you cannot set breakpoints in that code and yes, it would not break in that code unless to disable "just my code" in debug settings. Yes, I agree there should be better ways to debug external packages. But until there is, it is always best to directly reference a debug version of relevant dll to your project, otherwise you cannot debug it in any useful way.

**Reply by lostmsu** [reply]
@LTRData BTW, thank you for ImDisk - it was a sizable part of my development journeys!
